### ISSUE TITLE
[API Proposal]:  LoggerMessage - support ILogger property in addition to fields

### ISSUE THREAD

**Issue by abbotware** [dotnet/runtime#87747]
### Background and motivation

Add support for ILogger detection and use via a property in addition to field;

Motivation - allow the use of LoggerMessage without having to change existing base classes that use protected ILogger properties instead of fields.  By requiring the use of a non-private field, this pattern is in violation of:
- **CA1051:Do not declare visible instance fields**
- **SA1401: Field should be private**

Unless there is a general consensus that ILogger **properties** are a bad practice (if yes please provide relevant doc / links) this should be a straightforward enhancement.

### API Proposal

This function would need to be updated:

https://github.com/dotnet/dotnet/blob/cc80a55289b1069ffb69cdc891556cc5c4a6bfa8/src/runtime/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs#L411

as well as the messages:
https://github.com/dotnet/docs/blob/9564767df2b9623ac21a7b52a4596c36ad3e5f15/docs/fundamentals/syslib-diagnostics/syslib1019.md?plain=1#L9

and probably this:
https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator

### API Usage

no change to the usage - just support for more scenarios:

#### Example of Current Usage (violates CA1051 and SA1401)

```c#
public class BaseWithLoggerField
{
	// this violates rule CA1051:Do not declare visible instance fields
	// this violates rule SA1401: Field should be private
	protected readonly ILogger logger;

	public BaseWithLoggerField(ILogger<BaseWithLoggerField> logger)
	{
		this.logger = logger;
	}
}

public partial class Derived : BaseWithLoggerField
{
	public Derived(ILogger<Derived> logger) : base(logger)
	{
	}

	[LoggerMessage(Level = LogLevel.Debug, Message = "Logged: {something}")]
	private partial void LogSomething(string something);
}
```

#### Example of Usage with API Modification

This usage does not violate CA1051 / SA1401 and would be backwards compatible with current implementations above

```c#
public class BaseWithLoggerProperty
{
	public BaseWithLoggerProperty(ILogger<BaseWithLoggerProperty> logger)
	{
		this.Logger = logger;
	}

	protected ILogger Logger { get; }
}

public partial class Derived : BaseWithLoggerProperty
{
	public Derived(ILogger<Derived> logger) : base(logger)
	{
	}

	// currently, this produces error:
	// SYSLIB1019: Couldn't find a field of type Microsoft.Extensions.Logging.ILogger in class
	[LoggerMessage(Level = LogLevel.Debug, Message = "Logged: {something}")]
	private partial void LogSomething(string something);
}
```

### Risks

None

**Comment by tarekgh** [comment]
@abbotware could you please talk more about the benefit of the proposed change on the outcome result of the generator?

**Comment by abbotware** [comment]
using a protected field violates: CA1051:Do not declare visible instance fields

**Comment by tarekgh** [comment]
> using a protected field violates: CA1051:Do not declare visible instance fields

To be clear, do you have code example trigger that in your app?

**Comment by abbotware** [comment]
@tarekgh - updated API Usage with an example

**Comment by tarekgh** [comment]
Thanks @abbotware. 
I am wondering why you can't write the code like:

```C#
public class BaseWithLoggerField
{
	private ILogger _logger;

	public BaseWithLoggerField(ILogger<BaseWithLoggerField> logger)
	{
		this. Logger = logger;
	}
	
	protected ILogger Logger { get => _logger;  set => _logger = value; }
```

I am inquiring about the possibility of adding support for the logger property. However, implementing this feature might introduce complications if the class declares both a field and a property. It would not be appropriate to disallow this or generate an error when both are present, as it is possible to have a legitimate use case for such a scenario without necessarily knowing the intention behind the property implementation.

**Comment by abbotware** [comment]
@tarekgh -  that does nothing to address the problem: As the class name implies, it is **Base** class.  The problem I am running into is not in the base class, but the derived classes:

```
public partial class Derived : BaseWithLoggerField
{
	public Derived(ILogger<Derived> logger) : base(logger)
	{
	}

	// this does not work given your example:
	[LoggerMessage(Level = LogLevel.Debug, Message = "Logged: {something}")]
	private partial void LogSomething(string something);
}
```

**Comment by tarekgh** [comment]
@abbotware you don't have access to the base class? I was suggesting updating the base class which will make the derived class work fine. no?

**Comment by abbotware** [comment]
@tarekgh  - I do have access, but what you are proposing does not work- did you try to add a [LoggerMessage] method in a derived class?

**Comment by tarekgh** [comment]
The possible workaround here is to define the logger as internal in the base class. I know this is not the best, but it should work.
```
        internal ILogger _logger;
```

the other idea is just disabling this warning in the base class 

```
        #pragma warning disable CA1051
        protected ILogger _logger;
        #pragma warning restore CA1051
```

This is reasonable to do. We'll keep this issue open to see what that is best can be done here.

**Comment by abbotware** [comment]
No, it is **not reasonable**: You are telling me to **redesign** my base classes - this is a potentially **breaking** change.

The entire point of this proposal was to avoid doing exactly what you are suggesting.

**Comment by tarekgh** [comment]
> No, it is not reasonable: You are telling me to redesign my base classes - this is a potentially breaking change.

Why is this the case if you add ` #pragma warning disable CA1051`? nothing to redesign and not even a breaking change. This is what I meant by reasonable for now. I am keeping the issue to find a better solution.

**Comment by davidfowl** [comment]
@tarekgh if you don't own the base class (its in a library that you can't change for e.g.), then that approach is problematic.

**Comment by tarekgh** [comment]
Understand, my suggestion was specifically for @abbotware case who own the base class too. as I mentioned I am keeping this issue open to find a solution that work for everyone.

**Comment by cincuranet** [comment]
Triage: We think this is actually a feature gap we have and we should address it. Although this has low demand and hence we'll consider it as a low priority, especially given that this is going to be breaking change.
